mysys/crc32/crc32_arm64.c

```diff
@@ -68,250 +74,194 @@ asm(".arch_extension crypto");
+              :[v]"r"(*((const uint64_t *)buffer)));
+      buffer+= sizeof(uint64_t);
+    }
+# endif /* HAVE_ARMV8_CRC_CRYPTO_INTRINSICS */
+
+    /* Done if Input data size is aligned with 1024  */
+    if(!(length+= 1024))
+      return (~crc);
 
 #endif /* HAVE_ARMV8_CRYPTO */
 
-		while ((length -= sizeof(uint64_t)) >= 0) {
-			CRC32CX(crc, *(uint64_t *)buffer);
-			buffer += sizeof(uint64_t);
-		}
-		/* The following is more efficient than the straight loop */
-		if (length & sizeof(uint32_t)) {
-			CRC32CW(crc, *(uint32_t *)buffer);
-			buffer += sizeof(uint32_t);
-		}
-		if (length & sizeof(uint16_t)) {
-			CRC32CH(crc, *(uint16_t *)buffer);
-			buffer += sizeof(uint16_t);
-		}
-		if (length & sizeof(uint8_t))
-			CRC32CB(crc, *buffer);
-
-	} else {
-#ifdef HAVE_ARMV8_CRYPTO
-#ifdef HAVE_ARMV8_CRC_CRYPTO_INTRINSICS
-		const poly64_t k1 = 0xe417f38a;
-		uint64_t t0;
-		while ((length -= 1024) >= 0) {
-			crc0 = __crc32cd(crc, 0);
-
-			CRC32C7X3X8_ZERO;
-			CRC32C7X3X8_ZERO;
-			CRC32C7X3X8_ZERO;
-			CRC32C7X3X8_ZERO;
-			CRC32C7X3X8_ZERO;
-			CRC32C7X3X8_ZERO;
-
-			/* Merge crc0 into crc: crc0 multiply by K1 */
-			t0 = (uint64_t)vmull_p64(crc0, k1);
-			crc = __crc32cd(0, t0);
-		}
-#else /* !HAVE_ARMV8_CRC_CRYPTO_INTRINSICS */
-		__asm__("mov    x16,            #0xf38a         \n\t"
-			"movk   x16,            #0xe417, lsl 16 \n\t"
-			"mov    v1.2d[0],       x16             \n\t"
-			:::"x16");
-
-		while ((length -= 1024) >= 0) {
-			__asm__("crc32cx %w[c0], %w[c], xzr\n\t"
-				:[c0]"=r"(crc0):[c]"r"(crc));
-
-			CRC32C7X3X8_ZERO;
-			CRC32C7X3X8_ZERO;
-			CRC32C7X3X8_ZERO;
-			CRC32C7X3X8_ZERO;
-			CRC32C7X3X8_ZERO;
-			CRC32C7X3X8_ZERO;
-
-			__asm__("mov            v3.2d[0],       %x[c0]          \n\t"
-				"pmull          v3.1q,          v3.1d,  v1.1d   \n\t"
-				"mov            %x[c0],         v3.2d[0]        \n\t"
-				"crc32cx        %w[c],          wzr,    %x[c0]  \n\t"
-				:[c]"=r"(crc)
-				:[c0]"r"(crc0));
-		}
-#endif /* HAVE_ARMV8_CRC_CRYPTO_INTRINSICS */
-		if(!(length += 1024))
-			return (~crc);
-#endif /* HAVE_ARMV8_CRYPTO */
-		while ((length -= sizeof(uint64_t)) >= 0)
-			CRC32CX(crc, 0);
+  }  // end if pmull_supported
 
-		/* The following is more efficient than the straight loop */
-		if (length & sizeof(uint32_t))
-			CRC32CW(crc, 0);
+  while ((length-= sizeof(uint64_t)) >= 0)
+  {
+    CRC32CX(crc, *(uint64_t *)buffer);
+    buffer+= sizeof(uint64_t);
+  }
 
-		if (length & sizeof(uint16_t))
-			CRC32CH(crc, 0);
+  /* The following is more efficient than the straight loop */
+  if (length & sizeof(uint32_t))
+  {
+    CRC32CW(crc, *(uint32_t *)buffer);
+    buffer+= sizeof(uint32_t);
+  }
+
+  if (length & sizeof(uint16_t))
+  {
+    CRC32CH(crc, *(uint16_t *)buffer);
+    buffer+= sizeof(uint16_t);
+  }
 
-		if (length & sizeof(uint8_t))
-			CRC32CB(crc, 0);
-	}
+  if (length & sizeof(uint8_t))
+    CRC32CB(crc, *buffer);
 
-	return (~crc);
+  return (~crc);
```

Comment:
You could remove the redundant parentheses. (They were mandatory in InnoDB coding style a long time ago.)
```c
  return ~crc;
```